compiler/src/diagnostics/lenses.re

```diff
@@ -2,76 +2,154 @@ open Grain_typed;
+  make_value(~location, ~valType, ~sigStr);
+};
 
-        | _ => None
-        };
+let create_record_signature = (labels: list(Types.record_field)) => {
+  let recSig =
+    List.fold_left(
+      (acc, field: Types.record_field) => {
+        let comma = acc == "" ? "" : ", ";
+        let buf = Buffer.create(64);
+        let ppf = Format.formatter_of_buffer(buf);
+        Printtyp.type_expr(ppf, field.rf_type);
+        Format.pp_print_flush(ppf, ());
+        acc
+        ++ comma
+        ++ Ident.name(field.rf_name)
+        ++ " : "
+        ++ Buffer.contents(buf);
       },
-      program.statements,
+      "",
+      labels,
     );
+  "{" ++ recSig ++ " }";
+};
 
-  List.fold_left(
-    (acc, st: option(lens_t)) =>
-      switch (st) {
-      | None => acc
-      | Some(s) => List.append(acc, [s])
-      },
-    [],
-    lenses,
+let data_hover_val =
+    (
+      ~location: Grain_parsing.Location.t,
+      ~t: Types.type_declaration,
+      ~valType: string,
+    ) => {
+  let sigStr =
+    switch (t.type_kind) {
+    | TDataVariant(cstrs) => "Variant"
+    | TDataRecord(labels) => create_record_signature(labels)
+    | TDataAbstract => "Abstract"
+    | TDataOpen => "Open"
+    };
+
+  make_value(~location, ~valType, ~sigStr);
+};
+
+let print_tree = (stmts: list(Grain_typed.Typedtree.toplevel_stmt)) => {
+  let lenses = ref([]);
+
+  module Iterator =
+    TypedtreeIter.MakeIterator({
+      include TypedtreeIter.DefaultIteratorArgument;
+      let enter_expression = (exp: Grain_typed__Typedtree.expression) => {
+        let lens =
+          hover_val(~t=exp.exp_type, ~location=exp.exp_loc, ~valType="E");
+        lenses := List.append(lenses^, [lens]);
+      };
+      let enter_pattern = (pat: Grain_typed__Typedtree.pattern) => {
+        let lens =
+          hover_val(~t=pat.pat_type, ~location=pat.pat_loc, ~valType="P");
+        lenses := List.append(lenses^, [lens]);
+      };
+
+      let enter_data_declaration =
+          (d: Grain_typed__Typedtree.data_declaration) => {
+        let lens =
+          data_hover_val(~t=d.data_type, ~location=d.data_loc, ~valType="D");
+        lenses := List.append(lenses^, [lens]);
+      };
+    });
+
+  List.iter(
+    (cur: Grain_typed__Typedtree.toplevel_stmt) => {
+      let (file, startline, startchar, sbol) =
+        get_raw_pos_info(cur.ttop_loc.loc_start);
+      let (_, endline, endchar, ebol) =
+        get_raw_pos_info(cur.ttop_loc.loc_end);
+      let lens: lens_t = {
+        sl: startline,
+        sc: startchar,
+        sb: sbol,
+        el: endline,
+        ec: endchar,
+        eb: ebol,
+        s: "",
+        t: "S",
+      };
+
+      lenses := List.append(lenses^, [lens]);
+      switch (cur.ttop_desc) {
+      | TTopException(_)
+      | TTopForeign(_)
+      | TTopImport(_)
+      | TTopExport(_)
+      | TTopExpr(_)
+      | TTopLet(_) => Iterator.iter_toplevel_stmt(cur)
+      // this is separate as the original emitted an enter data
```

Comment:
What does this mean?

Reply:
Where I copied this code from, it emitted values for entering and leaving a TTopData .  I don't need it, so wanted to check with you, I'm going to simplify this switch away, I can always revert to it if the data is needed in the future.

Reply:
Gotcha!